### ISSUE TITLE
Run monorepo test suite against state resulting from proposal

### ISSUE THREAD

**Issue by mds1** [ethereum-optimism/superchain-ops#343]
This is discussed in https://github.com/ethereum-optimism/superchain-ops/issues/140#issuecomment-2075435988 through the use of [forge-proposal-simulator](https://github.com/solidity-labs-io/forge-proposal-simulator) (FPS). How to best do this needs some thought, and we should make sure our approach is forward compatible with batched upgrades (i.e. currently superchain-ops is focused on upgrading one task at a time but soon we will want to upgrade L1 contracts of many chains in 1 transaction—in that case we want to run the monorepo tests for each chain).

This is a big task, and likely will require changes both to this repo and the monorepo. We should check out how FPS does this, consider using that to avoid rolling our own solution, and may @ElliotFriedman will be kind enough to lend some time to help us out here :)

**Comment by mds1** [comment]
Yes definitely! So I recall last time we chatted about this a tricky part was that the upgrade transactions are executed in this repo, but our test suite is in the monorepo (https://github.com/ethereum-optimism/optimism/tree/develop/packages/contracts-bedrock). However this repo does already have the monorepo as a [dependency](https://github.com/ethereum-optimism/superchain-ops/blob/808004cd933e0bfa36e6d9124701e296639a6d41/.gitmodules#L7-L10), so that might help

First, some background on the monorepo setup:
- The core test suite is a standard forge test suite. We also have a few kontrol proofs, and a few differential ffi tests to ensure go and solidity implementations match.
- We run all tests (including the go ffi tests) with `just test`, and the proofs are run with `just test-kontrol`.
- Only running `just test` would be be a great start and kontrol can be added later.

Now on this repo, superchain-ops
- "Tasks"are transactions to broadcast, like contract upgrade transactions or transactions that just change a config value. This is essentially the same as what Maker calls "spells"
- All tasks live in a directory of `tasks/{L1-network}/{taskName}`, and contain `input.json` bundles that define the transaction to be sent, along with a single solidity script that simulates the task based on that JSON bundle. 
- Some tasks are "single" meaning it's a regular safe that needs to send the transaction, like the 5/7 Foundation Safe. Other tasks are "nested" meaning it's a 2/2 Safe that needs to send the transaction (where each signer on that 2/2 is a Safe, specifically the Foundation and the Security Council)
- The core logic to parse the bundle and simulate the transaction happens using the https://github.com/base-org/contracts/ dependency. We run the simulation in foundry within that dep, and have a `postCheck` hook that runs a bunch of assertions on the post-simulation state

I would say it's safe to consider most options on the table. Some examples I thought about:
- Does changing the shape of the `input.json` bundles help?
- Does replacing `base-org/contracts` with FPS help?
- Maker executes tests against their proposals like [this](https://github.com/makerdao/spells-mainnet/blob/master/src/DssSpell.t.sol), by basically defining a static test suite that always runs. This could also be an option for us, i.e. parameterize the key tests in the monorepo, that way we can import them here and run them

**Comment by blmalone** [comment]
After some consideration, we've landed on one possible solution to deliver this feature. 

At a high-level, we want to be able to grab the resultant state from the task that ran and then use that state inside the monorepos tests.

To do this, we can use a combination of foundry cheatcodes. Firstly, we can use [`dumpState(string calldata pathToStateJson)`](https://github.com/foundry-rs/forge-std/blob/master/src/Vm.sol#L1858). This produces a JSON file that contains allocs created during the duration of the script. e.g. 

```json
{
    "0x5aadfb43ef8daf45dd80f4676345b7676f1d70e3": {
        "nonce": "0x1",
        "balance": "0x0",
        "code": "0x68.......",
        "storage": {
            "0x0000000000000000000000000000000000000000000000000000000000000000": "0x0000000000000000000000000000000000000000000000000000000000000001",
            "0x0000000000000000000000000000000000000000000000000000000000000001": "0x68656c6c6f00000000000000000000000000000000000000000000000000000a"
        }
    }
}
```

With these allocs, we can _then_ use another foundry cheat code, namely [`loadAllocs(string calldata pathToAllocsJson)`](https://github.com/foundry-rs/forge-std/blob/master/src/Vm.sol#L1875). This loads in everything in the json file (_addresses, code, nonces, balances, state etc_). e.g. 

```solidity
vm.loadAllocs("../state3.json");
console.log(foo.val()); // would've reverted if loadAllocs had not have been invoked.
```
A note worth mentioning is that before we load allocations using the JSON file, we may want to do some processing to the file. As we know, deployments won't occur in superchain-ops tasks, so we may be able to remove any allocs that contain a `code` property, for example e.g. `MultisigTask` or `AddressRegistry`. 


---
Each test in the monorepo has the following structure: 
![Image](https://github.com/user-attachments/assets/1143d4b2-3f22-40c8-bf07-898f949a05ee)

We will want to integrate our new logic into the above architecture. We don't want our new logic to live within [`Deploy.s.sol`](https://github.com/ethereum-optimism/optimism/blob/de3865c0a2d32821faa19165ff1a3232589b90c1/packages/contracts-bedrock/scripts/deploy/Deploy.s.sol#L4) because we _don't_ want to deploy a new chain (proxies, implementation contracts etc). The reason being that all superchain-ops tasks work on existing chain states and in _most_ (if not all) cases, they don't deploy contracts as part of the task (think setting a new gas limit on an already existing SystemConfig contract). 

Our new logic will want to leverage [`ForkLive.s.sol`](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/test/setup/ForkLive.s.sol). Currently, `ForkLive.s.sol` reads addresses for a given network (mainnet/sepolia) from the superchain-registry and stores then so that they can be accessed later. This only grabs addresses for one chain at a time (e.g. base, zora or op), we may want to update this logic or loop at a higher level for all the chains in the superchain-ops task. 

We will need to update many files, not limited to: 
- https://github.com/ethereum-optimism/superchain-ops/blob/main/single.just
- https://github.com/ethereum-optimism/superchain-ops/blob/main/nested.just
- https://github.com/ethereum-optimism/superchain-ops/blob/main/.circleci/config.yml
- https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/justfile#L72
- https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/test/setup/ForkLive.s.sol
- https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/test/setup/Setup.sol

**Comment by blmalone** [comment]
Once we dump the state we will need to do some processing on it. We know this because as part of the `MultisigTask` scripts, we deploy various contracts. They include the `AddressRegistry.sol` and the script contract itself. In the _actual_ execution of the task, we know that these contracts are **not** deployed. Therefore, removing them is important to keep the allocs from the simulation as close to the production allocs as possible. 

In order to correctly identify the allocs we need to remove from the dumped state, we want to make sure that the deployed contracts have a deterministic address. 
We have a few options to realize this. The leading approach is to use [create3](https://github.com/0xsequence/create3) ([example](https://github.com/ethereum-optimism/optimism/blob/47c00f9975d2455d52263c0e54d52eb4b483b288/packages/contracts-bedrock/src/L2/OptimismSuperchainERC20Factory.sol#L60) of this being used in the monorepo). This is give us a deterministic address of the contracts that we deploy during a task. For the script contract itself, it's likely that this address is always the same and derived from a default forge address and nonce.

**Comment by blmalone** [comment]
https://github.com/ethereum-optimism/superchain-ops/issues/343#issuecomment-2610867210 
We're going to punt on state processing. For now, we don't know if we need this so we will delay it for now.

**Comment by mds1** [comment]
For performance reasons I would suggest only running the monorepo tests in CI on some schedule (e.g. every 4 hours) to avoid having to wait multiple minutes to simulate and execute tasks (e.g. if we ran tests as part of the simulate command)

**Comment by blmalone** [comment]
Closed by: https://github.com/ethereum-optimism/superchain-ops/pull/555
